backend/routes/auth.js

```diff
@@ -0,0 +1,147 @@
+const express = require("express");
+const { check, validationResult } = require("express-validator/check");
+const bcrypt = require("bcryptjs");
+const jwt = require("jsonwebtoken");
+const router = express.Router();
+
+const { User } = require("../models/user");
+
+/**
+ * @method - POST
+ * @param - /signup
+ * @description - User SignUp
+ */
+
+router.post(
+  "/signup",
+  [
+    check("username", "Please Enter a Valid Username").not().isEmpty(),
+    // check("email", "Please enter a valid email").isEmail(),
+    check("password", "Please enter a valid password").isLength({
+      min: 1,
+    }),
+  ],
+  async (req, res) => {
+    const errors = validationResult(req);
+    if (!errors.isEmpty()) {
+      return res.status(400).json({
+        errors: errors.array(),
+      });
+    }
+
+    const { username, password } = req.body;
+    try {
+      let user = await User.findOne({
+        username,
+      });
+      if (user) {
+        return res.status(400).json({
+          message: "User Already Exists",
+        });
+      }
+
+      user = new User({
+        username,
+        password,
+      });
+
+      const salt = await bcrypt.genSalt(10);
+      user.password = await bcrypt.hash(password, salt);
+
+      await user.save();
+
+      const payload = {
+        user: {
+          id: user.id,
+        },
+      };
+
+      jwt.sign(
+        payload,
+        "randomString", // change this to a stronger secret and don't commit it
+        {
+          expiresIn: 10000,
+        },
+        (err, token) => {
+          if (err) throw err;
+
+          res.status(200).json({
+            token,
+          });
+        }
+      );
+    } catch (err) {
+      console.log(err.message);
+      res.status(500).send("Error in Saving");
+    }
+  }
+);
+
+/**
+ * @method - POST
+ * @param - /login
+ * @description - User Login
+ */
+router.post(
+  "/login",
+  [
+    check("username", "Please Enter a Valid Username").not().isEmpty(),
+    //check("email", "Please enter a valid email").isEmail(),
+    check("password", "Please enter a valid password").isLength({
+      min: 1,
+    }),
+  ],
+  async (req, res) => {
+    const errors = validationResult(req);
+
+    if (!errors.isEmpty()) {
+      return res.status(400).json({
+        errors: errors.array(),
+      });
+    }
+
+    const { username, password } = req.body;
+    try {
+      let user = await User.findOne({
+        username,
+      });
+      if (!user)
+        return res.status(400).json({
+          message: "User does not exist",
```

Comment:
Usually when logging in and the user enters an incorrect username or password, the error message is just something like: "incorrect username or password", the user is not told which one is incorrect. This is done for security reasons to make it harder for a person to know which field was incorrect and randomly try guessing the 1 incorrect field. Can we change our code to match this? Even if just one of the fields is incorrect, let's just have 1 generic message that says: "Username of password was incorrect".

Reply:
Yea good point!